content/api/commands/attachfile.md

```diff
@@ -0,0 +1,213 @@
+
+### Attaching a file with custom filename and lastModified
+
+```javascript
+cy.get('input[type=file]').attachFile({
+  contents: 'path/to/file.json',
+  fileName: 'users.json',
+  lastModified: new Date('Feb 18 1989').valueOf()
+})
+```
+
+### Dropping a file on the document
+
+```javascript
+cy.document().attachFile('file.json', { action: 'drag-n-drop' })
+```
+
+## Notes
+
+### Existence
+
+#### Default file existence assertion
+
+When passed a file path, `cy.attachFile()` asserts that the file exists and will fail if it does not exist. It will retry reading the file if it does not initially exist until the file exists or the command times out.
+
+```javascript
+// will fail after the defaultCommandTimeout is reached
+cy.get('input[type=file]').attachFile('does-not-exist.yaml')
+```
+
+### Actionability
+
+#### The element must first reach actionability
+
+`.attachFile()` is an "action command" that follows all the rules
+[defined here](/guides/core-concepts/interacting-with-elements).
+
+## Rules
+
+### Requirements [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Chains-of-Commands)
+
+<List><li>`.attachFile()` requires being chained off a command that yields DOM
+element(s). With the `input` action (default), it further requires a single `input` element with `type="file"`, or a `label` element attached to one.</li><li>If given a path, `cy.attachFile()` requires the file must exist.</li><li>If given an alias, `cy.attachFile()` requires that the subject of the alias must not be `null` or `undefined`</li></List>
+
+### Assertions [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Assertions)
+
+<List><li>`.attachFile()` will automatically wait for the element to reach an
+[actionable state](/guides/core-concepts/interacting-with-elements)</li></List>
+
+### Timeouts [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Timeouts)
+
+<List><li>`.attachFile()` can time out waiting for the element to reach an
+[actionable state](/guides/core-concepts/interacting-with-elements).</li></List>
+
+## Command Log
+
+**_Attach file to input_**
+
+```javascript
+cy.get('.file-input').attachFile(Buffer.from('Hello world'))
+```
+
+The commands above will display in the Command Log as:
+
+<DocsImage src="/img/api/attachfile/attach-file-during-test.png" alt="Command log for attachFile" ></DocsImage>
+
+When clicking on `click` within the command log, the console outputs the
+following:
+
+<DocsImage src="/img/api/click/attach-file-in-console.png" alt="console.log for attachFile" ></DocsImage>
+
+## History
+
+| Version                                     | Changes                                                  |
+| ------------------------------------------- | -------------------------------------------------------- |
+| [10.0.0](/guides/references/changelog#10.0.0) | `cy.attachFile()` command added                        |
```

Comment:
Does this have to go into 10.0.0? I feel like this can go in an earlier release. Usually we wait to add the History table until we have the release branch.

```suggestion
| [10.0.0](/guides/references/changelog#10.0.0) | `cy.attachFile()` command added                        |
```

Reply:
So the reason I was thinking 10.0 is because of the popularity of the existing community solution, and the fact that adding a command that shares a name will throw an error if the plugin is still installed. While we did discuss 'not every fix is breaking' in standup today, and I agree it can sometimes be murky, but I think this one should be in a major version.

cypress-upload-file has [762 thousand](https://www.npmjs.com/package/cypress-file-upload) (!!!) weekly downloads - that's more a quarter as many downloads as Cypress itself. When we add in our own version of this command, every single one of those user's tests will break and they'll need to follow a migration guide.

It may theoretically be a new feature (minor release), but in practice it'll require a large portion of our users to update their code.

Reply:
If this is going into 10.0.0, it would be nice to see it based off of the `unification` branch. Otherwise, carry on!

Reply:
Switched to targeting the unification branch, thanks. I was looking for a 10.0-release branch matching the name of the cypress branch, or the 9.2-release branch naming we have in this repo and didn't see it.